src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h

```diff
@@ -0,0 +1,168 @@
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2016 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef SHM_SHMAM_EAGER_FBOX_INIT_H_INCLUDED
+#define SHM_SHMAM_EAGER_FBOX_INIT_H_INCLUDED
+
+#include "fbox_types.h"
+
+extern MPIDI_SHMAM_eager_fbox_control_t MPIDI_SHMAM_eager_fbox_control_global;
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_SHMAM_eager_init
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHMAM_eager_init(int rank, int grank, int num_local)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    int i;
+
+    MPIDI_SHMAM_fastbox_t *fastboxes_p = NULL;
+
+    MPIR_CHKPMEM_DECL(3);
+
+    MPIDI_SHMAM_eager_fbox_control_global.num_seg = 1;
+    MPIDI_SHMAM_eager_fbox_control_global.my_rank = rank;
+    MPIDI_SHMAM_eager_fbox_control_global.my_grank = grank;
+    MPIDI_SHMAM_eager_fbox_control_global.num_local = num_local;
+
+    MPIDI_SHMAM_eager_fbox_control_global.last_polled_grank = 0;
+
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.seg,
+                        MPIDU_shm_seg_info_ptr_t,
+                        MPIDI_SHMAM_eager_fbox_control_global.num_seg *
+                        sizeof(MPIDU_shm_seg_info_t), mpi_errno, "mem_region segments");
+
+    /* Request fastboxes region */
+
+    mpi_errno =
+        MPIDU_shm_seg_alloc(num_local * num_local * sizeof(MPIDI_SHMAM_fastbox_t),
+                            (void **) &fastboxes_p);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Request shared collective barrier vars region */
+
+    mpi_errno = MPIDU_shm_seg_alloc(MAX(sizeof(MPIDU_shm_barrier_t), MPIDU_SHM_CACHE_LINE_LEN),
+                                    (void **) &MPIDI_SHMAM_eager_fbox_control_global.
+                                    barrier_region);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Actually allocate the segment and assign regions to the pointers */
+
+    mpi_errno = MPIDU_shm_seg_commit(&MPIDI_SHMAM_eager_fbox_control_global.memory,
+                                     &MPIDI_SHMAM_eager_fbox_control_global.barrier,
+                                     num_local, grank, 0, rank);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Allocate table of pointers to fastboxes */
+
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.out,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+
+#define MPIDI_SHMAM_MAILBOX_INDEX(sender, receiver) ((num_local) * (sender) + (receiver))
```

Comment:
Defining something in between a function is rather gross.  Can this be moved to the top?